### DISCUSSION TITLE
How is package loading handled?

### DISCUSSION THREAD

**Discussion by billdenney** [ropensci/targets#849] [Help]
I occasionally have hard-to-trace issues with package handling.  My guess is that it's not a `targets` issue, but it's a "Bill's understanding of `targets` package handling" issue.

I tend to write my `_targets.R` file with minimal `library()` calls at the top (only `targets()` and `tarchetypes()` or other packages like `tarchetypes()` which are used for target list building).  And, I put my package loading for specific targets into the `packages` argument to `tar_target()`.

Occasionally, I tend to find an interaction between targets where I will get a function that is not expected to respond as the main function.  (Typically, it will be a conflict between `dplyr::select()` and `MASS::select()`.)  My solution is to defensively program with the library name, replacing `select()` with `dplyr::select()`, usually.  And with that, everything tends to work.

But, I'm wondering if I'm missing something in how I should be handling packages within `targets`.  I tried (and failed) to make a reprex in the below.  That is part of why I think it's my understanding of package handling rather than the package handling itself.  I expected the `eg_dplyr` target below to load both "MASS" and "dplyr" (and I think that it does behind the scenes).

Overall, I think that `targets` is doing the right things around package loading where e.g. packages in the `library()` calls at the top are loaded first, and then packages in the `packages` argument to `tar_target()` are loaded after that.  And, package loading is target-specific (e.g. if you load a package via the `packages` argument), it won't be loaded for the next target run (unless specifically requested).

With all of that, I looked in the manual and several help files to see where package handling is discussed, and I couldn't find it.  Can you please point me to the right docs to let me learn how package loading is handled (or let me know if it's not documented)?

```r
library(targets)

mass_select_example <- function() {
  # From the MASS documentation
  names(longley)[1] <- "y"
  lm.ridge(y ~ ., longley)
  select(lm.ridge(y ~ ., longley,
                  lambda = seq(0,0.1,0.0001)))
}

dplyr_select_example <- function(x) {
  starwars %>% select(height)
}

list(
  tar_target(
    eg_mass,
    mass_select_example(),
    packages="MASS"
  ),
  tar_target(
    eg_dplyr,
    dplyr_select_example(eg_mass),
    packages="dplyr"
  )
)
```

**Comment by wlandau** [comment]
I like your solution of using `::`, that's usually the best way to go. Otherwise, you can control the order in which the default set of packages are loaded using the `packages` target option, which you can set with `tar_option_set()`. `tar_option_get("packages")` defaults to the packages currently loaded with `library()` or `require` in that point in `_targets.R`, and it is in turn the default value of `packages` in `tar_target()`. `tar_option_set(packages = ...)` has the added benefit of avoiding extra package loads for targets that do not need to run (and for other functions like `tar_outdated()` and `tar_visnetwork()`).

There is an rOpenSci Slack thread  about using the `conflicted` package with `targets`, and there is friction because `conflicted` sets a bunch of settings in its own environment which does not carry over to targets on HPC workers. One workaround for that is to set the `conflicted` configuration in a project-level `.Rprofile` file. Not everyone likes that, but it seems to work.

**Comment by billdenney** [comment]
Thanks.  I didn't know about the `conflicted` package.  It does seem handy, but I prefer the manual way for now.  I would find very interesting an option for static analysis that used conflicted and all the loaded packages to see what does or may cause an issue.   Something like `tar_conflicted()` that ran conflicted on all packages mentioned in a `_targets.R` file.  (That's not exactly a feature request so much as a brainstorm.)

Am I correct that package load order in `targets` is:

1. First anything loaded with `library()` in the `_targets.R` file.
2. Then, anything loaded in the `packages` argument of any dependency target, recursively,
3. Then, anything loaded in the `packages` argument of the current target.

(And, mix in there if you call `library()` within a target, but that seems like bad form.)

**Reply by wlandau** [reply]
Well, not exactly. Each target has the package load step in (3). The default `packages` argument is `tar_option_get("packages")`, and the default `tar_option_get("packages")` is `(.packages())` (the packages you load in `_targets.R`). Your `library()` calls in `_targets.R` are executed before the pipeline starts, and those particular packages will remain loaded for targets that run on the main process but not for [workers in other R processes](http://books.ropensci.org/targets/hpc.html).

**Reply by billdenney** [reply]
I'm not sure if I'm reading this correctly now.  And, when I look at the link to the [workers in other R processes](http://books.ropensci.org/targets/hpc.html), I'm less certain that I understand package loading.  And, I think that the answer for how packages are loaded may be "it depends".  (But, that doesn't sound like how you would code something @wlandau, so I'm guessing that I'm missing something, still.)

I'm going to try to summarize it correctly again to see if I understand now.

* Some packages are loaded in the main process that coordinates the workers, and some packages are loaded in the worker processes.
* The packages loaded in the main process are not generally the same as those loaded in the worker processes.

For the main process:

* Packages are loaded by normal R methods (typically calls to `library()`).
* There is no other `targets`-specific method for loading packages in the main process.

For the worker processes:

* The packages are loaded based on the `packages` argument to `tar_target()`.
    * The default `packages` argument is `targets::tar_option_get("packages")` which loads the same libraries as from the main process above, by default.
    * If `targets::tar_option_set(packages=c("my", "list", "of", "packages"))` is used to set the packages, then that set of packages used.  And, the set of packages from `targets::tar_option_set()` takes effect from where it occurs in the `_targets.R` script.  So, any targets created above `targets::tar_option_set()` will use the prior package settings.
    * But, if the `packages` argument is manually specified, this does not apply.
* If the `packages` argument to `tar_target()` is manually set, that list is used.

Is the above correct?  Is the answer different for transient vs persistent workers?

**Reply by wlandau** [reply]
The main piece here is the `packages` argument of `tar_target()`. For each target that runs, `targets` calls `library()` on each package listed (in the order listed). Here, each package loads into the R session where the target is about to run. This could be the main controlling R session in the case of `tar_make()`, or a remote worker in the case of `tar_make_clustermq()` or `tar_make_future()`.

`tar_make_future()` uses transient workers, so the package loads only affect the current target (if the `future` plan is anything other than `sequential`). But for `tar_make_clustermq()` a worker can run multiple targets. So in the case of both `tar_make()` and `tar_make_clustermq()`, package loads from upstream packages could affect the loaded package environments of downstream packages.

I don't see this as a problem in the vast majority of cases. In fact, package loads usually add up to a noticeable amount of runtime, and it is usually best if a target can make use of packages already loaded. In addition, I cannot trust the process of unloading a package because package developers have complete control of `.onLoad()` and `.onAttach()`.

The other source of package loads is your own calls to `library()`, which could be in `_targets.R` (which affects the default `tar_option_get("packages")` or in a target (which applies to whatever R process the target is running in).

**Reply by billdenney** [reply]
@wlandau , Thanks!  Now I've got it!

My underlying issues were related to the fact that sometimes I was having issues with workers running multiple targets.  And, in some targets (seemingly randomly from my end), the packages were loaded from previous targets (causing issues with MASS vs dplyr `select()` being chosen).

Knowing all of that, does it make sense to alter the namespace search order when running a target so that the user-selected packages are at the top of the `search()` path in the expected order?  (Maybe changing search order the breaks things in bad ways.)

**Reply by wlandau** [reply]
I would rather not try to modify the search order. On top of the discussion at https://github.com/tidyverse/tidyverse/issues/159, to change the search path, it looks like `targets` would have to detach and reload many packages, which would be slow.

**Reply by billdenney** [reply]
That's fair.
